src/adapter/src/coord/timestamp_oracle.rs

```diff
@@ -186,4 +222,99 @@ mod tests {
 
         Ok(())
     }
+
+    pub async fn shareable_timestamp_oracle_impl_test<F, NewFn>(
+        mut new_fn: NewFn,
+    ) -> Result<(), anyhow::Error>
+    where
+        F: Future<Output = Arc<dyn ShareableTimestampOracle<Timestamp> + Send + Sync>>,
+        NewFn: FnMut(String, NowFn, Timestamp) -> F,
+    {
+        // Normally, these could all be separate test methods but we bundle them
+        // all together so that it's easier to call this one test method from
+        // the implementation tests.
+
+        // Timestamp::MIN as initial timestamp
+        let timeline = uuid::Uuid::new_v4().to_string();
+        let oracle = new_fn(timeline, NowFn::from(|| 0u64), Timestamp::MIN).await;
+        assert_eq!(oracle.read_ts().await, Timestamp::MIN);
+        assert_eq!(oracle.peek_write_ts().await, Timestamp::MIN);
+
+        // Timestamp::MAX as initial timestamp
+        let timeline = uuid::Uuid::new_v4().to_string();
+        let oracle = new_fn(timeline, NowFn::from(|| 0u64), Timestamp::MAX).await;
+        assert_eq!(oracle.read_ts().await, Timestamp::MAX);
+        assert_eq!(oracle.peek_write_ts().await, Timestamp::MAX);
+
+        // Timestamp::MAX-1 from NowFn. We have to step back by one, otherwise
+        // `write_ts` can't determine the "advance_to" timestamp.
+        let timeline = uuid::Uuid::new_v4().to_string();
+        let oracle = new_fn(
+            timeline,
+            NowFn::from(|| Timestamp::MAX.step_back().expect("known to work").into()),
+            Timestamp::MIN,
+        )
+        .await;
+        // At first, read_ts and peek_write_ts stay where they are.
+        assert_eq!(oracle.read_ts().await, Timestamp::MIN);
+        assert_eq!(oracle.peek_write_ts().await, Timestamp::MIN);
+        assert_eq!(
+            oracle.write_ts().await.timestamp,
+            Timestamp::MAX.step_back().expect("known to work")
+        );
+        // Now peek_write_ts jump to MAX-1 but read_ts stays.
+        assert_eq!(oracle.read_ts().await, Timestamp::MIN);
+        assert_eq!(
+            oracle.peek_write_ts().await,
+            Timestamp::MAX.step_back().expect("known to work")
+        );
```

Comment:
did you mean an `apply_write(MAX-1)` or an `apply_write(0)` here? If the latter, the `read_ts` will not actually advance to `MAX-1` because `apply_write` doesn't involve the NowFn.

Reply:
I meant `apply_write(Timestamp::MAX.step_back())`. It's not that important of a case to cover though.